distribution/pull_v2.go

```diff
@@ -877,6 +881,30 @@ func (p *v2Puller) pullSchema2Config(ctx context.Context, dgst digest.Digest) (c
 	return configJSON, nil
 }
 
+func retry(ctx context.Context, attempts int, sleep time.Duration, f func(ctx context.Context) error) (err error) {
+	for i := 0; i < attempts; i++ {
+		err = retryOnError(f(ctx))
+		if err == nil {
+			return nil
+		}
+		if _, isDNR := err.(xfer.DoNotRetry); isDNR {
+			logrus.WithError(err).Errorf("download failed after %d attempts", attempts)
+			return err
+		}
+
+		timer := time.NewTimer(sleep)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return ctx.Err()
+		case <-timer.C:
+			logrus.WithError(err).Error("retrying after error")
+			sleep *= 2
+		}
+	}
+	return fmt.Errorf("download failed after %d attempts: %w", attempts, err)
```

Comment:
See my other comment w.r.t. using `errors.Wrap()` and `attempts+1`

```suggestion
	return errors.Wrapf(err, "download failed after %d attempts", attempts+1)
```